File: core/pipeline.py
from __future__ import annotations

import json
import uuid
from typing import Any

from core.config import Config
from core.diffing import is_likely_mismatch, unified_diff
from core.prompts import load_template, render, split_system_user
from core.providers import Provider
from core.providers.base import ProviderError
from core.voice_profile import VoiceProfile


class PipelineError(Exception):
    pass


def humanize(
    provider: Provider,
    config: Config,
    voice_profile: VoiceProfile,
    draft: str,
    platform: str,
) -> dict[str, Any]:
    if not draft.strip():
        raise PipelineError("Draft text is empty.")

    used_voice_profile = not voice_profile.is_empty()
    voice_section = (
        voice_profile.as_prompt_context()
        if used_voice_profile
        else "No voice profile yet — apply only the generic AI-tell stripping above."
    )

    template = load_template("humanize.md")
    system, user_template = split_system_user(template)
    system = render(system, voice_section=voice_section, platform=platform or "Other")
    user = render(user_template, draft=draft)

    try:
        output = provider.complete(system, user, config.max_tokens)
    except ProviderError as exc:
        raise PipelineError(str(exc)) from exc

    if not output.strip():
        raise PipelineError("The provider returned an empty response.")

    output = output.strip()

    first_pass_draft = None
    if used_voice_profile:
        first_pass_draft = output
        second_user = render(user_template, draft=output)
        output = _second_pass(provider, config, system, second_user, fallback=output)

    return {
        "draft": output,
        "used_voice_profile": used_voice_profile,
        "first_pass_draft": first_pass_draft,
    }


def _second_pass(
    provider: Provider, config: Config, system: str, user: str, fallback: str
) -> str:
    """Run the humanize prompt again on its own output, so the model gets a
    second honest attempt at applying voice rules it missed the first time.
    Falls back to the first-pass draft if this call fails."""
    try:
        output = provider.complete(system, user, config.max_tokens)
    except ProviderError:
        return fallback

    output = output.strip()
    return output if output else fallback


def review(
    provider: Provider,
    config: Config,
    voice_profile: VoiceProfile,
    humanized_text: str,
    edited_text: str,
) -> dict[str, Any]:
    humanized_text = humanized_text.strip()
    edited_text = edited_text.strip()

    if not humanized_text or not edited_text:
        raise PipelineError("Both the humanized draft and the edited version are required.")

    if is_likely_mismatch(humanized_text, edited_text, config.mismatch_threshold):
        return {
            "mismatch": True,
            "message": "This looks like a different article, not an edit — check your paste.",
        }

    classification = _classify(provider, config, voice_profile, humanized_text, edited_text)

    style_count = sum(
        1 for c in classification.get("changes", []) if c.get("classification") == "style"
    )
    content_count = sum(
        1 for c in classification.get("changes", []) if c.get("classification") == "content"
    )

    rule_updates = classification.get("rule_updates", [])
    article_id = uuid.uuid4().hex[:8]
    actions = voice_profile.apply_rule_updates(rule_updates, article_id)
    voice_profile.save()

    new_rules = [a for a in actions if a["action"] == "new"]
    reinforced_rules = [a for a in actions if a["action"] == "reinforced"]

    merges: list[dict[str, Any]] = []
    if len(voice_profile.data.get("rules", [])) >= 2:
        try:
            merges = _consolidate(provider, config, voice_profile)
        except PipelineError:
            # Auto-consolidation is a nice-to-have on top of a successful review;
            # if the extra LLM call fails, keep the review result rather than
            # losing the already-saved rule updates.
            pass

    return {
        "mismatch": False,
        "changes": classification.get("changes", []),
        "style_count": style_count,
        "content_count": content_count,
        "edits_absorbed": len(actions),
        "summary": (
            f"{style_count} style edit{'s' if style_count != 1 else ''} "
            f"(added to voice profile), {content_count} content edit"
            f"{'s' if content_count != 1 else ''} (ignored for learning)."
        ),
        "profile_changes": {
            "new_rules": new_rules,
            "reinforced_rules": reinforced_rules,
            "merges": merges,
        },
    }


def consolidate_voice_profile(
    provider: Provider, config: Config, voice_profile: VoiceProfile
) -> dict[str, Any]:
    """Merge near-duplicate rules (same tendency, different wording) via an LLM pass.

    Public entry point for the manual "Consolidate Similar Rules" UI action.
    """
    rules = voice_profile.data.get("rules", [])
    before_count = len(rules)
    merges = _consolidate(provider, config, voice_profile)
    after_count = len(voice_profile.data.get("rules", []))
    return {
        "before_count": before_count,
        "after_count": after_count,
        "changed": after_count != before_count,
        "merges": merges,
    }


def _consolidate(
    provider: Provider, config: Config, voice_profile: VoiceProfile
) -> list[dict[str, Any]]:
    """Run the consolidation LLM pass and save. Returns the list of real merges."""
    rules = voice_profile.data.get("rules", [])
    if len(rules) < 2:
        return []

    template = load_template("consolidate.md")
    system, user_template = split_system_user(template)
    user = render(user_template, rules_json=voice_profile.rules_as_json())

    try:
        raw_output = provider.complete(system, user, config.max_tokens)
    except ProviderError as exc:
        raise PipelineError(str(exc)) from exc

    parsed = _parse_json_response(raw_output, "consolidation")
    clusters = parsed.get("clusters", []) if isinstance(parsed, dict) else []
    _after_count, merges = voice_profile.replace_with_clusters(clusters)
    voice_profile.save()
    return merges


def _classify(
    provider: Provider,
    config: Config,
    voice_profile: VoiceProfile,
    original: str,
    edited: str,
) -> dict[str, Any]:
    diff_text = unified_diff(original, edited)
    template = load_template("classify.md")
    system, user_template = split_system_user(template)
    user = render(
        user_template,
        original=original,
        edited=edited,
        diff=diff_text,
        existing_rules=voice_profile.rules_context_for_matching(),
    )

    try:
        raw_output = provider.complete(system, user, config.max_tokens)
    except ProviderError as exc:
        raise PipelineError(str(exc)) from exc

    parsed = _parse_json_response(raw_output, "classifier")
    parsed.setdefault("changes", [])
    parsed.setdefault("rule_updates", [])
    return parsed


def _parse_json_response(raw_output: str, label: str) -> dict[str, Any]:
    text = raw_output.strip()
    if text.startswith("```"):
        text = text.strip("`")
        if text.startswith("json"):
            text = text[4:]
        text = text.strip()

    try:
        parsed = json.loads(text)
    except json.JSONDecodeError as exc:
        raise PipelineError(
            f"The {label} response could not be parsed as JSON. Try again."
        ) from exc

    if not isinstance(parsed, dict):
        raise PipelineError(f"The {label} response was not a JSON object.")

    return parsed

File: macos-app/Sources/HumanizerApp/Pipeline.swift
import Foundation

// MARK: - LLM JSON response shapes

struct ClassifyChange: Codable {
    var originalSnippet: String?
    var editedSnippet: String?
    var classification: String?
    var reasoning: String?

    enum CodingKeys: String, CodingKey {
        case classification, reasoning
        case originalSnippet = "original_snippet"
        case editedSnippet = "edited_snippet"
    }
}

struct ClassifyRuleUpdate: Codable {
    var matchedRuleID: String?
    var description: String?
    var exampleBefore: String?
    var exampleAfter: String?
    var category: String?

    enum CodingKeys: String, CodingKey {
        case description, category
        case matchedRuleID = "matched_rule_id"
        case exampleBefore = "example_before"
        case exampleAfter = "example_after"
    }
}

struct ClassifyResponse: Codable {
    var changes: [ClassifyChange] = []
    var ruleUpdates: [ClassifyRuleUpdate] = []

    enum CodingKeys: String, CodingKey {
        case changes
        case ruleUpdates = "rule_updates"
    }
}

struct ConsolidateCluster: Codable {
    var description: String?
    var category: String?
    var exampleBefore: String?
    var exampleAfter: String?
    var sourceRuleIDs: [String]?

    enum CodingKeys: String, CodingKey {
        case description, category
        case exampleBefore = "example_before"
        case exampleAfter = "example_after"
        case sourceRuleIDs = "source_rule_ids"
    }
}

struct ConsolidateResponse: Codable {
    var clusters: [ConsolidateCluster] = []
}

// MARK: - Results surfaced to the UI

struct HumanizeResult {
    var draft: String
    var usedVoiceProfile: Bool
    /// The output of the first (humanize) pass, before the verify pass ran —
    /// nil if the verify pass didn't run. Surfaced in the UI for debugging
    /// which pass introduced a given change.
    var preVerifyDraft: String?
}

struct ReviewResult {
    var mismatch: Bool
    var message: String?
    var changes: [ClassifyChange]
    var styleCount: Int
    var contentCount: Int
    var editsAbsorbed: Int
    var summary: String
    var newRules: [RuleUpdateAction]
    var reinforcedRules: [RuleUpdateAction]
    var merges: [MergeResult]
}

struct ConsolidateResult {
    var beforeCount: Int
    var afterCount: Int
    var changed: Bool
    var merges: [MergeResult]
}

@MainActor
enum Pipeline {
    static func humanize(provider: Provider, maxTokens: Int, voiceProfile: VoiceProfileStore, draft: String, platform: String) async throws -> HumanizeResult {
        guard !draft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw PipelineError.message("Draft text is empty.")
        }

        let usedVoiceProfile = !voiceProfile.isEmpty
        let voiceSection = usedVoiceProfile
            ? voiceProfile.asPromptContext()
            : "No voice profile yet — apply only the generic AI-tell stripping above."

        let system = PromptTemplates.render(PromptTemplates.humanizeSystem, [
            "voice_section": voiceSection,
            "platform": platform.isEmpty ? "Other" : platform,
        ])
        let user = PromptTemplates.render(PromptTemplates.humanizeUser, ["draft": draft])

        var output: String
        do {
            output = try await provider.complete(system: system, user: user, maxTokens: maxTokens)
        } catch let error as ProviderError {
            throw PipelineError.message(error.errorDescription ?? "Provider error.")
        }

        guard !output.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw PipelineError.message("The provider returned an empty response.")
        }
        output = output.trimmingCharacters(in: .whitespacesAndNewlines)

        let preVerifyDraft = output
        let mandatoryRules = voiceProfile.mandatoryRulesContext()
        if !mandatoryRules.isEmpty {
            output = await verify(provider: provider, maxTokens: maxTokens, mandatoryRules: mandatoryRules, draft: output)
        }

        return HumanizeResult(
            draft: output,
            usedVoiceProfile: usedVoiceProfile,
            preVerifyDraft: mandatoryRules.isEmpty ? nil : preVerifyDraft
        )
    }

    /// Second pass: check draft against mandatory (high-confidence) rules and fix any
    /// missed instances. Falls back to the unverified draft if this call fails, since a
    /// failed proofreading pass shouldn't block humanize.
    private static func verify(provider: Provider, maxTokens: Int, mandatoryRules: String, draft: String) async -> String {
        let system = PromptTemplates.render(PromptTemplates.verifySystem, ["mandatory_rules": mandatoryRules])
        let user = PromptTemplates.render(PromptTemplates.verifyUser, ["draft": draft])

        guard let output = try? await provider.complete(system: system, user: user, maxTokens: maxTokens) else {
            return draft
        }
        let trimmed = output.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? draft : trimmed
    }

    static func review(provider: Provider, maxTokens: Int, mismatchThreshold: Double, voiceProfile: VoiceProfileStore, humanizedText: String, editedText: String) async throws -> ReviewResult {
        let humanized = humanizedText.trimmingCharacters(in: .whitespacesAndNewlines)
        let edited = editedText.trimmingCharacters(in: .whitespacesAndNewlines)

        guard !humanized.isEmpty, !edited.isEmpty else {
            throw PipelineError.message("Both the humanized draft and the edited version are required.")
        }

        if Diffing.isLikelyMismatch(humanized, edited, threshold: mismatchThreshold) {
            return ReviewResult(
                mismatch: true,
                message: "This looks like a different article, not an edit — check your paste.",
                changes: [], styleCount: 0, contentCount: 0, editsAbsorbed: 0,
                summary: "", newRules: [], reinforcedRules: [], merges: []
            )
        }

        let classification = try await classify(provider: provider, maxTokens: maxTokens, voiceProfile: voiceProfile, original: humanized, edited: edited)

        let styleCount = classification.changes.filter { $0.classification == "style" }.count
        let contentCount = classification.changes.filter { $0.classification == "content" }.count

        let articleID = UUID().uuidString.prefix(8).lowercased()
        let actions = voiceProfile.applyRuleUpdates(classification.ruleUpdates, articleID: String(articleID))
        voiceProfile.save()

        let newRules = actions.filter { $0.action == "new" }
        let reinforcedRules = actions.filter { $0.action == "reinforced" }

        var merges: [MergeResult] = []
        if voiceProfile.data.rules.count >= 2 {
            merges = (try? await consolidateInternal(provider: provider, maxTokens: maxTokens, voiceProfile: voiceProfile)) ?? []
        }

        let summary = "\(styleCount) style edit\(styleCount != 1 ? "s" : "") (added to voice profile), " +
            "\(contentCount) content edit\(contentCount != 1 ? "s" : "") (ignored for learning)."

        return ReviewResult(
            mismatch: false,
            message: nil,
            changes: classification.changes,
            styleCount: styleCount,
            contentCount: contentCount,
            editsAbsorbed: actions.count,
            summary: summary,
            newRules: newRules,
            reinforcedRules: reinforcedRules,
            merges: merges
        )
    }

    static func consolidateVoiceProfile(provider: Provider, maxTokens: Int, voiceProfile: VoiceProfileStore) async throws -> ConsolidateResult {
        let beforeCount = voiceProfile.data.rules.count
        let merges = try await consolidateInternal(provider: provider, maxTokens: maxTokens, voiceProfile: voiceProfile)
        let afterCount = voiceProfile.data.rules.count
        return ConsolidateResult(beforeCount: beforeCount, afterCount: afterCount, changed: afterCount != beforeCount, merges: merges)
    }

    private static func consolidateInternal(provider: Provider, maxTokens: Int, voiceProfile: VoiceProfileStore) async throws -> [MergeResult] {
        guard voiceProfile.data.rules.count >= 2 else { return [] }

        let system = PromptTemplates.consolidateSystem
        let user = PromptTemplates.render(PromptTemplates.consolidateUser, ["rules_json": voiceProfile.rulesAsJSON()])

        let raw: String
        do {
            raw = try await provider.complete(system: system, user: user, maxTokens: maxTokens)
        } catch let error as ProviderError {
            throw PipelineError.message(error.errorDescription ?? "Provider error.")
        }

        let parsed: ConsolidateResponse = try parseJSONResponse(raw, label: "consolidation")
        let (_, merges) = voiceProfile.replaceWithClusters(parsed.clusters)
        voiceProfile.save()
        return merges
    }

    private static func classify(provider: Provider, maxTokens: Int, voiceProfile: VoiceProfileStore, original: String, edited: String) async throws -> ClassifyResponse {
        let diffText = Diffing.unifiedDiff(original, edited)
        let system = PromptTemplates.classifySystem
        let user = PromptTemplates.render(PromptTemplates.classifyUser, [
            "original": original,
            "edited": edited,
            "diff": diffText,
            "existing_rules": voiceProfile.rulesContextForMatching(),
        ])

        let raw: String
        do {
            raw = try await provider.complete(system: system, user: user, maxTokens: maxTokens)
        } catch let error as ProviderError {
            throw PipelineError.message(error.errorDescription ?? "Provider error.")
        }

        return try parseJSONResponse(raw, label: "classifier")
    }

    private static func parseJSONResponse<T: Decodable>(_ raw: String, label: String) throws -> T {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.hasPrefix("```") {
            text = text.trimmingCharacters(in: CharacterSet(charactersIn: "`"))
            if text.hasPrefix("json") {
                text = String(text.dropFirst(4))
            }
            text = text.trimmingCharacters(in: .whitespacesAndNewlines)
        }

        guard let data = text.data(using: .utf8) else {
            throw PipelineError.message("The \(label) response could not be parsed as JSON. Try again.")
        }
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw PipelineError.message("The \(label) response could not be parsed as JSON. Try again.")
        }
    }
}
